File: Typescript_bootcamp/Group_Anagrams.ts
function groupAnagrams(strs: string[]): string[][]
{
    let cache: Record<string, string[]> = {};

    strs.forEach((str) => {
        const sortedString = str.split('').sort().join('');
        cache[sortedString] = cache[sortedString] ?? [];
        cache[sortedString].push(str);
    })

    return Object.values(cache);
};

// console.log(groupAnagrams(["eat","tea","tan","ate","nat","bat"]));

File: Python_bootcamp/Hash_Table/Group_Anagrams.py
from typing import List
from collections import defaultdict
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        groups = defaultdict(list)
        
        for s in strs:
            sorted_s = ''.join(sorted(s))
            groups[sorted_s].append(s)
            
        return list(groups.values())
    
strs: List[str] = ["eat", "tea", "tan", "ate", "nat", "bat"]
print(Solution.groupAnagrams(strs))